compiler/test/runnable/exe2.c

```diff
@@ -0,0 +1,1405 @@
+            i = (int *)(-(long)j);
+        }
+}
+
+/* The following is exe8.c with alignment       */
+
+#define bug(text,name,p) {printf("\n%s name (%%p) = %p\n",("text"),(name));}
+
+typedef struct {
+    char var1[10];
+    struct {
+        char var2[2];
+        char var3[3];
+        char var4[4];
+        char var5[5];
+        char var6;
+        } inner;
+} TYPE1;
+
+struct {
+    char var1[10];
+    struct {
+        char var2[2];
+        char var3[3];
+        char var4[4];
+#if c_plusplus
+        typedef const volatile int asdkfj(const i);
+#endif
+        char var5[5];
+        char var6;
+        } inner;
+} str2;
+
+TYPE1 buf1, *buf1p = &buf1;
+TYPE1 arr[3];
+
+typedef union
+{       char var[7];
+} TYPE2;
+
+TYPE2 buf2, *buf2p = &buf2;
+
+void alignment()                /* always compiled without -a */
+{
+        TYPE1 *buf1p2;
+        TYPE2 *buf2p2;
+
+        bug(should be 26:,(void*)sizeof(TYPE1),d)
+#if __INTSIZE == 4
+        printf("TYPE1 = %d\n",sizeof(TYPE1));
+        assert(sizeof(TYPE1) == 25);
+#else
+        assert(sizeof(TYPE1) == 25);
+#endif
+
+        buf1p2 = buf1p + 1;
+        assert(buf1p2 == (TYPE1 *) ((char *) buf1p + sizeof(TYPE1)));
+        buf1p2 = buf1p + sizeof(TYPE1);
+        assert((buf1p2 - buf1p) == sizeof(TYPE1));
+
+        buf2p2 = buf2p + 1;
+        assert(buf2p2 == (TYPE2 *) ((char *) buf2p + sizeof(TYPE2)));
+        buf2p2 = buf2p + sizeof(TYPE2);
+        assert((buf2p2 - buf2p) == sizeof(TYPE2));
+
+        bug(should be 1:,(void*)sizeof(buf1p->inner.var6),d)
+        assert(sizeof(buf1p->inner.var6) == 1);
+
+        bug(should be 10:,(void*)sizeof(buf1p->var1),d)
+        assert(sizeof(buf1p->var1) == 10);
+        bug(should be 1:,(void*)sizeof(buf1.var1[2]),d)
+        assert(sizeof(buf1.var1[2]) == 1);
+        bug(should be 10:,(void*)sizeof(str2.var1),d)
+        assert(sizeof(str2.var1) == 10);
+
+        bug(should be 5:,(void*)sizeof(buf1p->inner.var5),d)
+        assert(sizeof(buf1p->inner.var5) == 5);
+        bug(should be 1:,(void*)sizeof(*buf1.inner.var5),d)
+        assert(sizeof(*buf1.inner.var5) == 1);
+        bug(should be 5:,(void*)sizeof(str2.inner.var5),d)
+        assert(sizeof(str2.inner.var5) == 5);
+
+        /* ideally, this should be true to match hardware structs */
+        bug(should be 78 when compiled without -a:,(void*)sizeof(arr),d)
+        /*printf("%d\n",sizeof(arr));*/
+#if __INTSIZE == 4
+        assert(sizeof(arr) == 75);
+#else
+        assert(sizeof(arr) == 75);
+#endif
+}
+
+void preprocessor()
+{       int i = 0;
+#if 1
+        i = 3;
+#elif 0
+        assert(0);
+#else
+        assert(0);
+#endif
+        assert(i == 3);
+
+#if 0
+        assert(0);
+#elif 1
+        i = 4;
+#else
+        assert(0);
+#endif
+        assert(i == 4);
+
+#if 0
+        assert(0);
+#elif 0
+        assert(0);
+#else
+        i = 5;
+#endif
+        assert(i == 5);
+}
+
+/********** COMPILE-ONLY CODE *****************/
+
+typedef struct linechg_
```

Comment:
```diff
--- a/compiler/test/runnable/exe2.c
+++ b/compiler/test/runnable/exe2.c
@@ -897,7 +897,7 @@ void preprocessor()
 #endif
         assert(i == 5);
 }
-
+
 /********** COMPILE-ONLY CODE *****************/
 ```
some weird non standard whitespace here?

Reply:
Oh, there was a form feed character. I used to use them long ago :-)

Reply:
Ah, that explains it.